<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="generator" content="pandoc">
  <meta name="author" content="Peter Ralph">
  <title>Scaling up genomic computation with the tree sequence</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="./reveal.js/dist/reset.css">
  <link rel="stylesheet" href="./reveal.js/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    span.underline{text-decoration: underline;}
    div.column{display: inline-block; vertical-align: top; width: 50%;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
  </style>
  <link rel="stylesheet" href="./reveal.js/dist/theme/simple.css" id="theme">
  <style type="text/css">

  .reveal { font-size: 30px; }

  .reveal h1 { font-size: 1.5em; } 

  .reveal h2 { font-size: 1.2em; } 

  .reveal .slides { text-align: left; }

  .reveal .slides figure { text-align: center; }

  .centered p { text-align: center; }

  /* apply this to the columns div */
  div.vcentered { display: flex; align-items: center; }

  .caption p { font-size: 75%; }

  .greyed { color: #4444; }

  div.floatright { float: right; text-align: right; }

  .reveal figcaption { display: none; }

  .reveal img { border: none; }

  .reveal section img { border: none; }

  .MathJax_Display { margin: 0; }

  img.finger { vertical-align: text-top; }

  .smaller { font-size: 50% }
  .smallish { font-size: 80% }

  </style>

  <div style="display: none">
  \[
  \newcommand{\E}{\mathbb{E}}
  \renewcommand{\P}{\mathbb{P}}
  \DeclareMathOperator{\var}{var}
  \]
  </div>
</head>
<body>
  <div class="reveal">
    <div class="slides">

<section id="title-slide">
  <h1 class="title">Scaling up genomic computation with the tree sequence</h1>
  <p class="author">Peter Ralph</p>
  <p class="date">Department of Statistics <br/> University of Michigan // 18 Feb 2022</p>
</section>

<section>
<section id="outline" class="title-slide slide level1">
<h1>Outline</h1>

</section>
<section id="outline-of-the-talk" class="slide level2">
<h2>Outline of the talk</h2>
<ol type="1">
<li>Genomes, and simulations</li>
<li>The tree sequence</li>
<li>Applications</li>
</ol>
<p><em>slides:</em> <a href="https://petrelharp.github.io/umich-feb-2022/umich-feb-2022.slides.html">github.com/petrelharp/umich-feb-2022</a></p>
</section></section>
<section>
<section id="genomes-and-genealogies" class="title-slide slide level1">
<h1>Genomes and genealogies</h1>

</section>
<section id="genomes" class="slide level2">
<h2>Genomes</h2>
<ul>
<li class="fragment">are very big (<span class="math inline">\(10^7\)</span>–<span class="math inline">\(10^{12}\)</span> nucleotides)</li>
<li class="fragment">encode the basic <em>mechanisms</em> of life</li>
<li class="fragment">reflect past <em>history</em> and <em>process</em></li>
</ul>
<div class="fragment">
<figure>
<img data-src="figs/kreitman-1983.png" alt="Genotype matrix from Kreitman 1983" /><figcaption aria-hidden="true">Genotype matrix from Kreitman 1983</figcaption>
</figure>
</div>
</section>
<section id="section" class="slide level2">
<h2></h2>
<p>Your two copies of the genome were inherited, noisily, from your two parents,</p>
<div class="fragment">
<p>and from your four grandparents,</p>
</div>
<div class="fragment">
<p>and your eight great-grandparents</p>
</div>
<div class="fragment">
<p>and your sixteen great-great-grandparents</p>
</div>
</section>
<section id="genealogical-ancestors" class="slide level2">
<h2>Genealogical ancestors</h2>
<p>In a well-mixed population of size <span class="math inline">\(N\)</span>,</p>
<ul>
<li><p>A most recent common ancestor of <em>everyone</em> lived at <span class="math inline">\(\approx \log_2 N\)</span> generations ago.</p></li>
<li><p>For humans: <span class="math inline">\(\approx 500\)</span> years; but including geography and history, <span class="math inline">\(\approx\)</span> 2–3 thousand years (Chang ’99; Rohde, Olson &amp; Chang ’04)</p></li>
<li><p><em>Everyone</em> alive <span class="math inline">\(\approx 1.77 \log_2 N\)</span> generations ago is either an ancestor of everyone or has left behind no descendants. (Chang ’99)</p></li>
</ul>
</section>
<section id="meiosis-recombination" class="slide level2">
<h2>Meiosis &amp; Recombination</h2>
<div class="columns">
<div class="column" style="width:50%;">
<figure>
<img data-src="figs/recombination-cartoon-crop.png" alt="recombination" /><figcaption aria-hidden="true">recombination</figcaption>
</figure>
</div><div class="column" style="width:50%;">
<ul>
<li><p>You have two copies of each chromosome, one from each parent.</p></li>
<li><p>When you make a gamete, the copies <em>recombine</em>, at Poisson(<span class="math inline">\(\rho\)</span>) locations.</p></li>
<li><p><em>Mutations</em> appear at Poisson(<span class="math inline">\(\mu\)</span>) locations.</p></li>
</ul>
</div>
</div>
</section>
<section id="you-and-your-greatk-2-grand-parents" class="slide level2">
<h2>You and your great<span class="math inline">\({}^{k-2}\)</span>-grand parents</h2>
<p>The number of <span class="math inline">\(k^\text{th}\)</span> <em>genetic</em> ancestors grows <em>linearly</em> with <span class="math inline">\(k\)</span>.</p>
<div class="centered">
<p><img data-src="figs/gcbias/num_genetics_vs_genealogical_ancs.png" /></p>
</div>
<div class="caption">
<p>from <a href="https://gcbias.org/2013/11/11/how-does-your-number-of-genetic-ancestors-grow-back-over-time/">gcbias.org</a></p>
</div>
</section>
<section id="section-1" class="slide level2">
<h2></h2>
<div class="centered">
<p><img data-src="figs/gcbias/parents_grandparents.png" style="width:60.0%" /></p>
</div>
<div class="caption">
<p>from <a href="https://gcbias.org/2013/11/11/how-does-your-number-of-genetic-ancestors-grow-back-over-time/">gcbias.org</a></p>
</div>
</section>
<section id="section-2" class="slide level2">
<h2></h2>
<div class="centered">
<p><img data-src="figs/gcbias/male_line.png" style="width:60.0%" /></p>
</div>
<div class="caption">
<p>from <a href="https://gcbias.org/2013/11/11/how-does-your-number-of-genetic-ancestors-grow-back-over-time/">gcbias.org</a></p>
</div>
</section>
<section id="section-3" class="slide level2">
<h2></h2>
<p>Everyone is related by a vast genealogical network.</p>
<div class="fragment">
<p>We get to observe <em>genomes</em>, noisily transmitted through it.</p>
</div>
<div class="fragment">
<p>What can we learn about its structure? history? biology?</p>
</div>
</section></section>
<section>
<section id="simulation-based-inference" class="title-slide slide level1">
<h1>Simulation-based inference</h1>

</section>
<section id="some-questions" class="slide level2">
<h2>Some questions</h2>
<ol type="1">
<li><p>What forces contribute to the variation in genetic diversity along the genome? <em>(explaining variation in diversity)</em></p></li>
<li><p>Which locations along the genome have been the recent targets of positive natural selection? <em>(identifying sweeps)</em></p></li>
<li><p>Where did this individual come from? <em>(inferring location)</em></p></li>
<li><p>How do organisms disperse across the landscape? <em>(dispersal maps)</em></p></li>
</ol>
</section>
<section id="section-4" class="slide level2">
<h2></h2>
<p><img data-src="figs/modeling_reality.png" /></p>
</section>
<section id="section-5" class="slide level2">
<h2></h2>
<p><img data-src="figs/modeling_model_parameters.png" /></p>
</section>
<section id="section-6" class="slide level2">
<h2></h2>
<p><img data-src="figs/modeling_model_parameters_inverse.png" /></p>
</section>
<section id="section-7" class="slide level2">
<h2></h2>
<p><img data-src="figs/modeling_parameters_inverse_computer.png" /></p>
</section>
<section id="simulation-based-inference-1" class="slide level2">
<h2>Simulation-based inference</h2>
<div class="centered">
<p><img data-src="figs/modeling_parameters_inverse_computer.png" style="width:60.0%" /></p>
</div>
<ul>
<li>bespoke confirmatory simulations</li>
<li>optimization of one or two parameters <!-- *(if between-simulation noise is small)* --></li>
<li>Approximate Bayesian Computation (ABC)</li>
<li>deep learning</li>
</ul>
</section></section>
<section>
<section id="what-do-we-need" class="title-slide slide level1">
<h1>What do we need</h1>

</section>
<section id="section-8" class="slide level2">
<h2></h2>
<ul>
<li><p>Fast simulation of genomes</p></li>
<li><p>Fast computation of summary statistics</p></li>
</ul>
</section>
<section id="wish-list" class="slide level2">
<h2>Wish list:</h2>
<div class="smallish">
<div class="columns">
<div class="column">
<p>Whole genomes, thousands of samples,<br />
from millions of individuals.</p>
<p><strong>Demography:</strong></p>
<ul>
<li>life history</li>
<li>separate sexes</li>
<li>selfing</li>
<li>polyploidy</li>
<li>species interactions</li>
</ul>
<p><strong>Geography:</strong></p>
<ul>
<li>discrete populations</li>
<li>continuous landscapes</li>
<li>barriers</li>
</ul>
<p><strong>History:</strong></p>
<ul>
<li>ancient samples</li>
<li>range shifts</li>
</ul>
</div><div class="column">
<p><strong>Natural selection:</strong></p>
<ul>
<li>selective sweeps</li>
<li>introgressing alleles</li>
<li>background selection</li>
<li>quantitative traits</li>
<li>incompatibilities</li>
<li>local adaptation</li>
</ul>
<p><strong>Genomes:</strong></p>
<ul>
<li>recombination rate variation</li>
<li>gene conversion</li>
<li>infinite-sites mutation</li>
<li>nucleotide models</li>
<li>context-dependence</li>
<li>mobile elements</li>
<li>inversions</li>
<li>copy number variation</li>
</ul>
</div>
</div>
</div>
</section>
<section id="section-9" class="slide level2" data-background-image="figs/oregon_geological_map.png">
<h2 data-background-image="figs/oregon_geological_map.png"></h2>
</section>
<section id="enter-slim" class="slide level2">
<h2>Enter SLiM</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>by Ben Haller and Philipp Messer</p>
<p>an individual-based, scriptable forwards simulator</p>
<div class="caption">
<p><img data-src="figs/ben-haller.jpg" alt="Ben Haller" /> <em>Ben Haller</em></p>
</div>
</div><div class="column" style="width:50%;">
<figure>
<img data-src="figs/slim-gui.png" alt="SLiM GUI" /><figcaption aria-hidden="true">SLiM GUI</figcaption>
</figure>
<p><a href="https://messerlab.org/SLiM/">messerlab.org/SLiM</a></p>
</div>
</div>
</section>
<section id="section-10" class="slide level2" data-background-image="figs/slim_screenshot.png" data-background-position="center" data-background-size="100%">
<h2 data-background-image="figs/slim_screenshot.png" data-background-position="center" data-background-size="100%"></h2>
</section>
<section id="section-11" class="slide level2">
<h2></h2>
<div class="smallish">
<div class="columns">
<div class="column">
<ul>
<li><s>Whole genomes,</s>*</li>
<li><s>thousands of samples, </s></li>
<li><s>from millions of individuals.</s>*</li>
</ul>
<p><strong>Demography:</strong></p>
<ul>
<li><s>life history</s></li>
<li><s>separate sexes</s>*</li>
<li><s>selfing</s></li>
<li>polyploidy*</li>
<li>species interactions <strong>(coming soon!)</strong></li>
</ul>
<p><strong>Geography:</strong></p>
<ul>
<li><s>discrete populations</s></li>
<li><s>continuous landscapes</s></li>
<li><s>barriers</s>*</li>
</ul>
<p><strong>History:</strong></p>
<ul>
<li><s>ancient samples</s></li>
<li><s>range shifts</s></li>
</ul>
</div><div class="column">
<p><strong>Natural selection:</strong></p>
<ul>
<li><s>selective sweeps</s></li>
<li><s>introgressing alleles</s></li>
<li><s>background selection</s></li>
<li><s>quantitative traits</s>*</li>
<li><s>incompatibilities</s>*</li>
<li><s>local adaptation</s>*</li>
</ul>
<p><strong>Genomes:</strong></p>
<ul>
<li><s>recombination rate variation</s></li>
<li><s>gene conversion</s></li>
<li><s>infinite-sites mutation</s></li>
<li><s>nucleotide models</s></li>
<li><s>context-dependence</s>*</li>
<li>mobile elements*</li>
<li>inversions*</li>
<li>copy number variation</li>
</ul>
</div>
</div>
</div>
</section>
<section id="section-12" class="slide level2">
<h2></h2>
<ul>
<li><s>Whole genomes,</s>*</li>
</ul>
<div class="fragment">
<p>Idea: if we record <em>how everyone is related to everyone else</em>,</p>
</div>
<div class="fragment">
<p>we can put down neutral mutations <em>after the simulation is over</em> instead of carrying them along.</p>
</div>
<div class="fragment">
<p>Since neutral mutations don’t affect demography,</p>
<p>this is <em>equivalent</em> to having kept track of them throughout.</p>
<!-- Tree sequences -->
</div>
</section></section>
<section>
<section id="the-tree-sequence" class="title-slide slide level1">
<h1>The tree sequence</h1>

</section>
<section id="history-is-a-sequence-of-trees" class="slide level2">
<h2>History is a sequence of trees</h2>
<p>For a set of sampled chromosomes, at each position along the genome there is a genealogical tree that says how they are related.</p>
<div class="fragment">
<figure>
<img data-src="figs/example.svg" alt="Trees along a chromosome" /><figcaption aria-hidden="true">Trees along a chromosome</figcaption>
</figure>
</div>
</section>
<section class="slide level2">

<p>The <strong>succinct tree sequence</strong></p>
<div class="floatright">
<p>is a way to succinctly describe this, er, sequence of trees</p>
<p><em>and</em> the resulting genome sequences.</p>
<div class="caption">
<p><a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004842">Kelleher, Etheridge, &amp; McVean</a></p>
</div>
</div>
<div class="fragment">
<div class="columns">
<div class="column" style="width:50%;">
<figure>
<img data-src="figs/tskit_logo.png" style="width:80.0%" alt="tskit logo" /><figcaption aria-hidden="true">tskit logo</figcaption>
</figure>
</div><div class="column" style="width:50%;">
<div class="floatright">
<figure>
<img data-src="figs/jerome.jpeg" style="width:50.0%" alt="jerome kelleher" /><figcaption aria-hidden="true">jerome kelleher</figcaption>
</figure>
<div class="caption">
<p>jerome kelleher</p>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="example-three-samples-two-trees-two-variant-sites" class="slide level2">
<h2>Example: three samples; two trees; two variant sites</h2>
<figure>
<img data-src="figs/example_tree_sequence.png" alt="Example tree sequence" /><figcaption aria-hidden="true">Example tree sequence</figcaption>
</figure>
</section>
<section id="nodes-and-edges" class="slide level2">
<h2>Nodes and edges</h2>
<dl>
<dt>Edges</dt>
<dd><p>Who inherits from who.</p>
<p>Records: interval (left, right); parent node; child node.</p>
</dd>
<dt>Nodes</dt>
<dd><p>The ancestors those happen in.</p>
<p>Records: time ago (of birth); ID (implicit).</p>
</dd>
</dl>
</section>
<section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.0.png" alt="Building a tree sequence" /><figcaption aria-hidden="true">Building a tree sequence</figcaption>
</figure>
</section>
<section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.1.png" alt="Building a tree sequence" /><figcaption aria-hidden="true">Building a tree sequence</figcaption>
</figure>
</section>
<section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.2.png" alt="Building a tree sequence" /><figcaption aria-hidden="true">Building a tree sequence</figcaption>
</figure>
</section>
<section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.3.png" alt="Building a tree sequence" /><figcaption aria-hidden="true">Building a tree sequence</figcaption>
</figure>
</section>
<section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.4.png" alt="Building a tree sequence" /><figcaption aria-hidden="true">Building a tree sequence</figcaption>
</figure>
</section>
<section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.5.png" alt="Building a tree sequence" /><figcaption aria-hidden="true">Building a tree sequence</figcaption>
</figure>
</section>
<section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.6.png" alt="Building a tree sequence" /><figcaption aria-hidden="true">Building a tree sequence</figcaption>
</figure>
</section>
<section id="sites-and-mutations" class="slide level2">
<h2>Sites and mutations</h2>
<dl>
<dt>Mutations</dt>
<dd><p>When state changes along the tree.</p>
<p>Records: site it occured at; node it occurred in; derived state.</p>
</dd>
<dt>Sites</dt>
<dd><p>Where mutations fall on the genome.</p>
<p>Records: genomic position; ancestral (root) state; ID (implicit).</p>
</dd>
</dl>
</section>
<section class="slide level2">

<figure>
<img data-src="sites_muts_walkthrough/sites_muts_walkthrough.0.png" alt="Adding mutations" /><figcaption aria-hidden="true">Adding mutations</figcaption>
</figure>
</section>
<section class="slide level2">

<figure>
<img data-src="sites_muts_walkthrough/sites_muts_walkthrough.1.png" alt="Adding mutations" /><figcaption aria-hidden="true">Adding mutations</figcaption>
</figure>
</section>
<section class="slide level2">

<figure>
<img data-src="sites_muts_walkthrough/sites_muts_walkthrough.2.png" alt="Adding mutations" /><figcaption aria-hidden="true">Adding mutations</figcaption>
</figure>
</section>
<section class="slide level2">

<figure>
<img data-src="sites_muts_walkthrough/sites_muts_walkthrough.3.png" alt="Adding mutations" /><figcaption aria-hidden="true">Adding mutations</figcaption>
</figure>
</section>
<section class="slide level2">

<figure>
<img data-src="sites_muts_walkthrough/sites_muts_walkthrough.4.png" alt="Adding mutations" /><figcaption aria-hidden="true">Adding mutations</figcaption>
</figure>
</section>
<section class="slide level2">

<p><strong>The result:</strong> an encoding of the genomes <em>and</em> all the genealogical trees.</p>
<div class="centered">
<figure>
<img data-src="figs/example_tree_sequence.png" alt="Example tree sequence" /><figcaption aria-hidden="true">Example tree sequence</figcaption>
</figure>
</div>
</section></section>
<section>
<section id="hows-it-work" class="title-slide slide level1">
<h1>How’s it work?</h1>

</section>
<section id="file-sizes" class="slide level2">
<h2>File sizes</h2>
<div class="centered">
<figure>
<img data-src="figs/tsinfer_sizes.png" style="width:90.0%" alt="file sizes" /><figcaption aria-hidden="true">file sizes</figcaption>
</figure>
</div>
<div class="caption">
<p>100Mb chromosomes; from <a href="https://www.nature.com/articles/s41588-019-0483-y">Kelleher et al 2018, <em>Inferring whole-genome histories in large population datasets</em></a>, Nature Genetics</p>
</div>
<!-- Estimated sizes of files required to store the genetic variation data for a
simulated human-like chromosome (100 megabases) for up to 10 billion haploid
(5 billion diploid) samples. Simulations were run for 10 1 up to 10 7 haplotypes
using msprime [Kelleher et al., 2016], and the sizes of the resulting files plotted
(points). -->
</section>
<section class="slide level2">

<figure>
<img data-src="figs/ts_ex/tree_sequence_genotypes.png" alt="genotypes" /><figcaption aria-hidden="true">genotypes</figcaption>
</figure>
</section>
<section class="slide level2">

<figure>
<img data-src="figs/ts_ex/tree_sequence_genotype_and_tree.png" alt="genotypes and a tree" /><figcaption aria-hidden="true">genotypes and a tree</figcaption>
</figure>
</section>
<section class="slide level2">

<figure>
<img data-src="figs/ts_ex/tree_sequence_next_genotype_and_tree.png" alt="genotypes and the next tree" /><figcaption aria-hidden="true">genotypes and the next tree</figcaption>
</figure>
</section>
<section id="for-n-samples-genotyped-at-m-sites" class="slide level2">
<h2>For <span class="math inline">\(N\)</span> samples genotyped at <span class="math inline">\(M\)</span> sites</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><em>Genotype matrix</em>:</p>
<p><span class="math inline">\(N \times M\)</span> things.</p>
</div><div class="column" style="width:50%;">
<p><em>Tree sequence:</em></p>
<ul>
<li><span class="math inline">\(2N-2\)</span> edges for the first tree</li>
<li><span class="math inline">\(\sim 4\)</span> edges per each of <span class="math inline">\(T\)</span> trees</li>
<li><span class="math inline">\(M\)</span> mutations</li>
</ul>
<p><span class="math inline">\(O(N + T + M)\)</span> things</p>
</div>
</div>
<figure>
<img data-src="figs/ts_ex/tree_sequence_genotype_and_tree.png" style="width:60.0%" alt="genotypes and a tree" /><figcaption aria-hidden="true">genotypes and a tree</figcaption>
</figure>
</section></section>
<section>
<section id="summarizing-genomes-and-genealogies" class="title-slide slide level1">
<h1>Summarizing genomes and genealogies</h1>

</section>
<section id="fast-genotype-statistics" class="slide level2">
<h2>Fast genotype statistics</h2>
<div class="centered">
<figure>
<img data-src="figs/treestats/benchmarks_without_copy_longer_genome.png" style="width:70.0%" alt="efficiency of treestat computation" /><figcaption aria-hidden="true">efficiency of treestat computation</figcaption>
</figure>
</div>
<div class="caption">
<p>from <a href="https://www.biorxiv.org/content/10.1101/779132v1?rss=1">R., Thornton and Kelleher 2019, <em>Efficiently summarizing relationships in large samples</em></a>, bioRxiv</p>
</div>
</section>
<section id="summaries-of-genotypes-and-genealogies" class="slide level2">
<h2>Summaries of genotypes and genealogies</h2>
<div class="columns">
<div class="column" style="width:47%;">
<p><em>Genotypes:</em></p>
<ol type="1">
<li>For each site,</li>
<li>look at who has which alleles,</li>
<li>and add a <em>summary</em> of these values to our running total.</li>
</ol>
<p><em>Example:</em> genetic distance counts how many mutations differ between two sequences.</p>
</div><div class="column" style="width:5%;">

</div><div class="column" style="width:47%;">
<!-- nothing til next slide -->
</div>
</div>
</section>
<section id="summaries-of-genotypes-and-genealogies-1" class="slide level2">
<h2>Summaries of genotypes and genealogies</h2>
<div class="columns">
<div class="column" style="width:47%;">
<p><em>Genotypes:</em></p>
<ol type="1">
<li>For each site,</li>
<li>look at who has inherited which alleles,</li>
<li>and add a <em>summary</em> of these values to the running total.</li>
</ol>
<p><em>Example:</em> sequence divergence counts how many mutations differ between two sequences.</p>
</div><div class="column" style="width:5%;">

</div><div class="column" style="width:47%;">
<p><em>Trees:</em></p>
<ol type="1">
<li>For each branch,</li>
<li>look at who would inherit mutations on that branch,</li>
<li>and add the <em>expected contribution</em> to the running total.</li>
</ol>
<p><em>Example:</em> the mean time to most recent common ancestor between two sequences.</p>
</div>
</div>
</section>
<section id="section-13" class="slide level2">
<h2></h2>
<figure>
<img data-src="figs/ts_ex/tree_sequence_site_and_branch.png" alt="site and branch stats" /><figcaption aria-hidden="true">site and branch stats</figcaption>
</figure>
</section>
<section id="section-14" class="slide level2">
<h2></h2>
<p>Given</p>
<ol type="1">
<li>a <em>weight</em> <span class="math inline">\(w_i \in \mathbb{R}^n\)</span> to each <em>sample node</em>, and</li>
<li>a <em>summary function</em> <span class="math inline">\(f : \mathbb{R}^n \to \mathbb{R}\)</span>,</li>
</ol>
<div class="fragment">
<p>the <strong>Site</strong> statistic <span class="math display">\[\begin{equation}
 \text{Site}(f,w) = \sum_{i: \text{sites}} \sum_{a: \text{alleles}_i} f(w_a)
\end{equation}\]</span> is the total summarized weights of all <em>alleles</em> at all sites,</p>
</div>
<div class="fragment">
<p>and the <strong>Branch</strong> statistic <span class="math display">\[\begin{equation}
 \text{Branch}(f,w) = \sum_{T: \text{trees}} s_T \sum_{b: \text{branches}_T} f(w_b) \ell_b
\end{equation}\]</span> is the <em>expected value</em> of <span class="math inline">\(\text{Site}(f,w)\)</span> under Poisson(1) mutation, given the trees.</p>
</div>
</section>
<section id="section-15" class="slide level2">
<h2></h2>
<p>With genealogies <em>fixed</em>, and averaging only over <em>mutations</em> with rate <span class="math inline">\(\mu\)</span>, <span class="math display">\[\begin{equation}
    \text{Branch}(f, w) = \frac{1}{\mu} \E\left[ \text{Site}(f, w) \vert T \right] .
\end{equation}\]</span></p>
<div class="fragment">
<p>Dealing directly with genealogies can <em>remove the layer of noise due to mutation</em>: <span class="math display">\[\begin{equation}
    \frac{1}{\mu^2} \var\left[\text{Site}(f,w)\right]
    =
    \var\left[\text{Branch}(f,w)\right]
    +
    \frac{1}{n}
    \E\left[\text{Branch}(f^2,w)\right]
\end{equation}\]</span></p>
<div class="caption greyed floatright">
<p>also see R., TPB, 2019</p>
</div>
</div>
</section>
<section id="section-16" class="slide level2">
<h2></h2>
<div class="columns">
<div class="column" style="width:60%;">
<figure>
<img data-src="figs/treestats/relate_chr20_site_div_branch.1000000.diversity.png" style="width:100.0%" alt="duality in 1000G data" /><figcaption aria-hidden="true">duality in 1000G data</figcaption>
</figure>
</div><div class="column" style="width:40%;">
<p>Duality, on 1000 Genomes data? Not quite…</p>
<ul>
<li>variation in mutation rate?</li>
<li>biased gene conversion?</li>
<li>selection?</li>
<li>inference artifacts?</li>
</ul>
<div class="caption">
<p><em>Tree sequence from <a href="https://www.nature.com/articles/s41588-019-0484-x">Speidel et al 2019</a>.</em></p>
</div>
</div>
</div>
</section></section>
<section>
<section id="application-to-genomic-simulations" class="title-slide slide level1">
<h1>Application to genomic simulations</h1>

</section>
<section id="the-main-idea" class="slide level2">
<h2>The main idea</h2>
<p>If we <em>record the tree sequence</em> that relates everyone to everyone else,</p>
<p>after the simulation is over we can put neutral mutations down on the trees.</p>
<div class="fragment">
<p>Since neutral mutations don’t affect demography,</p>
<p>this is <em>equivalent</em> to having kept track of them throughout.</p>
</div>
<div class="fragment">
<div class="columns">
<div class="column" style="width:50%;">
<div class="caption">
<p>From Kelleher, Thornton, Ashander, and Ralph 2018, <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1006581">Efficient pedigree recording for fast population genetics simulation</a>.</p>
<p>and Haller, Galloway, Kelleher, Messer, and Ralph 2018, <a href="https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.12968"><em>Tree‐sequence recording in SLiM opens new horizons for forward‐time simulation of whole genomes</em></a></p>
</div>
</div><div class="column" style="width:50%;">
<div class="floatright">
<p><img data-src="figs/jared.jpeg" style="width:35.0%" alt="jared galloway" /> <img data-src="figs/jaime.jpg" style="width:30.0%" alt="jaime ashander" /> <img data-src="figs/ben-haller.jpg" style="width:30.0%" alt="ben haller" /></p>
</div>
</div>
</div>
</div>
</section>
<section class="slide level2">

<p>This means recording the entire genetic history of <strong>everyone</strong> in the population, <strong>ever</strong>.</p>
<div class="fragment">
<p>It is <em>not</em> clear this is a good idea.</p>
</div>
<div class="fragment">
<p>But, with a few tricks…</p>
</div>
</section>
<section id="section-17" class="slide level2">
<h2></h2>
<p>The key idea: simplify</p>
</section>
<section id="a-100x-speedup" class="slide level2">
<h2>A 100x speedup!</h2>
<div class="columns">
<div class="column" style="width:40%;">
<div class="centered">
<figure>
<img data-src="figs/slim_logo.png" style="width:100.0%" alt="SLiM logo" /><figcaption aria-hidden="true">SLiM logo</figcaption>
</figure>
</div>
</div><div class="column" style="width:60%;">
<div class="floatright">
<p><img data-src="figs/slim_timing_no_msprime.png" style="width:100.0%" /></p>
</div>
</div>
</div>
</section>
<section id="what-else-can-you-do-with-tree-sequences" class="slide level2">
<h2>What else can you do with tree sequences?</h2>
<ul>
<li class="fragment">recorded pedigree and migration history</li>
<li class="fragment">true ancestry assignment</li>
<li class="fragment"><em>recapitation</em>: fast, post-hoc initialization with coalescent simulation</li>
<li class="fragment">fast, convenient computation</li>
</ul>
</section>
<section class="slide level2">

<p><strong>For example:</strong></p>
<ul>
<li>genome as human chr7 (<span class="math inline">\(1.54 \times 10^8\)</span>bp)</li>
<li><span class="math inline">\(\approx\)</span> 10,000 diploids</li>
<li>500,000 overlapping generations</li>
<li>continuous, square habitat</li>
<li>selected mutations at rate <span class="math inline">\(10^{-10}\)</span></li>
<li>neutral mutations added afterwards</li>
</ul>
<p><em>Runtime:</em> 8 hours on a single core</p>
</section></section>
<section>
<section id="example-predicting-location" class="title-slide slide level1">
<h1>Example: predicting location</h1>
<p><img data-src="figs/spatial_pedigree.png" /></p>
<div class="caption">
<p>from <a href="https://arxiv.org/abs/1904.09847">Bradburd &amp; Ralph 2019</a></p>
</div>
</section>
<section id="section-18" class="slide level2">
<h2></h2>
<div class="centered">
<p><img data-src="figs/locator_methods.png" /></p>
<p><a href="https://elifesciences.org/articles/54507"><img data-src="figs/locator_paper.png" style="width:50.0%" /></a> <img data-src="figs/cjb.jpg" style="width:20.0%" /></p>
</div>
</section>
<section id="locator-battey-et-al-2020" class="slide level2">
<h2>locator (<a href="https://elifesciences.org/articles/54507">Battey et al 2020</a>)</h2>
<p><img data-src="figs/locator_results.png" /></p>
</section>
<section id="section-19" class="slide level2">
<h2></h2>
<p><img data-src="figs/locator_spatial_tree_sequence.png" /></p>
<div class="flushright">
<div class="caption">
<p>from <a href="https://elifesciences.org/articles/54507">Battey et al 2020</a></p>
</div>
</div>
</section></section>
<section>
<section id="example-dispersal-maps" class="title-slide slide level1">
<h1>Example: dispersal maps</h1>

</section>
<section class="slide level2">

<figure>
<img data-src="figs/spacetree_1.png" alt="trees in space, by CJ Battey" /><figcaption aria-hidden="true">trees in space, by CJ Battey</figcaption>
</figure>
<div class="caption">
<p>by <a href="cjbattey.com">CJ Battey</a></p>
</div>
</section>
<section class="slide level2">

<figure>
<img data-src="figs/everyone-ibd.png" alt="genetic and geographic distance for desert tortoises" /><figcaption aria-hidden="true">genetic and geographic distance for desert tortoises</figcaption>
</figure>
<ul>
<li>genetic versus geographic distance between pairs of 272 desert tortoises (McCartney-Melstad, Shaffer)</li>
<li>clouds are comparisons within/between the two colors</li>
</ul>
</section>
<section class="slide level2">

<p><img data-src="figs/range-abundance-map.jpeg" /></p>
</section>
<section class="slide level2">

<p><img data-src="figs/drecp-pref-alt-snapshot.png" /></p>
</section></section>
<section>
<section id="wrap-up" class="title-slide slide level1">
<h1>Wrap-up</h1>

</section>
<section id="software-development-goals" class="slide level2">
<h2>Software development goals</h2>
<div class="columns">
<div class="column" style="width:50%;">
<ul>
<li>open</li>
<li>welcoming and supportive</li>
<li>reproducible and well-tested</li>
<li>backwards compatible</li>
<li>well-documented</li>
<li>capacity building</li>
</ul>
<!--
::: {.centered}
![popsim logo](figs/popsim.png){width=50%}
:::
-->
</div><div class="column" style="width:50%;">
<div class="centered">
<figure>
<img data-src="figs/tskit_logo.png" style="width:60.0%" alt="tskit logo" /><figcaption aria-hidden="true">tskit logo</figcaption>
</figure>
<p><a href="https://tskit.dev">tskit.dev</a></p>
<!-- ![SLiM logo](figs/slim_logo.png){width=80%} -->
</div>
</div>
</div>
</section>
<section id="thanks" class="slide level2">
<h2>Thanks!</h2>
<div class="columns">
<div class="column" style="width:50%;">
<ul>
<li>Andy Kern</li>
<li>Matt Lukac</li>
<li>Murillo Rodrigues</li>
<li>Victoria Caudill</li>
<li>Anastasia Teterina</li>
<li>Gilia Patterson <!--
- Saurabh Belsare
- Chris Smith
- Gabby Coffing
- Jeff Adrion
--></li>
<li>CJ Battey</li>
<li>Jared Galloway</li>
<li>the rest of the Co-Lab</li>
</ul>
<p>Funding:</p>
<ul>
<li>NIH NIGMS</li>
<li>NSF DBI</li>
<li>Sloan foundation</li>
<li>UO Data Science</li>
</ul>
</div><div class="column" style="width:50%;">
<div style="font-size: 85%; margin-top: -40px;">
<ul>
<li>Jerome Kelleher</li>
<li>Ben Haller</li>
<li>Ben Jeffery</li>
<li>Georgia Tsambos</li>
<li>Jaime Ashander</li>
<li>Gideon Bradburd</li>
<li>Madeline Chase</li>
<li>Bill Cresko</li>
<li>Alison Etheridge</li>
<li>Evan McCartney-Melstad</li>
<li>Brad Shaffer</li>
<li>Sean Stankowski</li>
<li>Matt Streisfeld</li>
</ul>
</div>
<div class="floatright">
<p><img data-src="figs/tskit_logo.png" style="width:40.0%" alt="tskit logo" /> <img data-src="figs/slim_logo.png" style="width:40.0%" alt="SLiM logo" /></p>
</div>
</div>
</div>
</section>
<section id="section-20" class="slide level2" data-background-image="figs/guillemots_thanks.png" data-background-position="center" data-background-size="50%">
<h2 data-background-image="figs/guillemots_thanks.png" data-background-position="center" data-background-size="50%"></h2>
</section></section>
    </div>
  </div>

  <script src="./reveal.js/dist/reveal.js"></script>

  // reveal.js plugins
  <script src="./reveal.js/plugin/notes/notes.js"></script>
  <script src="./reveal.js/plugin/search/search.js"></script>
  <script src="./reveal.js/plugin/zoom/zoom.js"></script>
  <script src="./reveal.js/plugin/math/math.js"></script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
        // Display the page number of the current slide
        slideNumber: true,
        // Push each slide change to the browser history
        history: true,
        // Transition style
        transition: 'none', // none/fade/slide/convex/concave/zoom
        math: {
          mathjax: 'https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [
          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    </body>
</html>
